### PR TITLE
Batch of website in backend fixes: website edition with rtl languages, Editor Menu hidden on translatable website, document title without 'Odoo -' prefix

### PR DESCRIPTION
**Pull request by younn-o**
[FIX] website: remove 'Odoo' prefix on the WebsitePreview document title

Before this commit, the WebsitePreview document title was not completely
replaced with the iframe's one: it was still prefixed by 'Odoo - ' by
the title service.

Now, the WebsitePreview, introduced in [1], is adapted to remove the
'zopenerp' part when replacing the title (and adding it again when
unmounted). While doing this, the backend's favicon is also changed with
the frontend's one, to be consistent with displaying the frontend
document title only.

This commit also adds an effect on the Optimize SEO dialog, so that the
document title matches the user's input.

[1]: https://github.com/odoo/odoo/commit/31cc10b91dc7762e23b4bde9b945be0c4ce3fe3b

task-2687506

-----

[FIX] web_editor, website: fix website edition with rtl languages

Before this commit, for the following flow:
- Add arabic language (or any other rtl language),
- On one of the websites, set it as the only website's language,
- Reload and go to the client action,
=> The frontend is correctly displayed in rtl,
- Click on edit,
=> The frontend, in the iframe, is reverted to ltr,

The WysiwygAdapter, introduced in [1], was not passing the correct
direction option to the OdooEditor, which would revert the editable to
the default 'ltr' direction.

[1]: https://github.com/odoo/odoo/commit/31cc10b91dc7762e23b4bde9b945be0c4ce3fe3b

task-2687506

-----

[FIX] website: hide "Editor Menu" on translatable websites

Before this commit, the Editor Menu was confusing when opened for a
translatable website: it would list the menus in the user's language,
not in the displayed website's language.

First, this commit shows that menu only for websites in their default
language, as the user should edit his website structure in the default
language.
Secondly, the 'get_tree' and 'save' requests on the website menus are
done with the website's language, not the user's language.

This commit also fixes a bug introduced with [1]: the website override
of the start method of the SnippetsMenu was incorrectly done.
In translate mode, it should not activate snippets on clicks, otherwise,
the LinkPopover was instantiated when clicking on navbar menus, allowing
to edit the menus from the translate mode, and breaking it.

[1]: https://github.com/odoo/odoo/commit/df1869153a90898bad3e0b61e5fc43a9ed3d59c9

task-2687506

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website/static/src/components/dialog/edit_menu.js
Comment (rdeodoo) [inline addons/website/static/src/components/dialog/edit_menu.js:121]: Doesn't this need its counterpart in the save :grimacing: ?

File: addons/web_editor/controllers/main.py
Comment (rdeodoo) [inline addons/web_editor/controllers/main.py:511]: That's unusual to say the least.
First why is `kwargs` not `**kwargs`? That's probably something missed during httpocalypse, or I am misunderstanding something.
If those were `**kwargs`, context would simply be pop from it and this wouldn't be possible, which would be good.
I am not sure to understand how would the context passed be different than the request one, as passing lang in context kwargs in JS will make it impact the request env lang

So, isn't this a noop? I mean that `request.env['ir.ui.view']` is already equal to `View = request.env['ir.ui.view'].with_context(lang=lang)` no?

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Is it wanted that we now show the Menu Editor in translate mode following frontend > backend merge?

**Comment by rdeodoo** [comment]
I'd like to have @qsm-odoo and @pierre-pde opinion about this:
- Before frontend > backend merge, we didn't display the menu editor in translated language, but now we do
- It was reported in the pad that viewing the menu in english when we were in french was strange, this pr fixed that
- Now, it works fine, you can edit the menu in the lang you are, it will save as translation if you are in translated mode
- But something is still strange, it looks like I am just editing the translated menu, which is true for the wording, but not for the structure. If I move a menu below another, I am not doing that on the translated version only but on all versions. Same if I add a menu, it looks like I am editing the french (translated) menu and the new "XYZ" menu I create will be related to the french version but it is not, it is just a new entry in the menu in all languages.

If I have to chose, I'd actually restore what we had before, make the Menu Editor only visible in the main language
And if not, I'd not make it fetch the translated terms but keep it as it is right now (without this PR) and just always load the main language version, it should be more clear than you are editing the structure / main one and not the translated version.. Unless nothing is yet translated, you might be thinking that you are editing the translations, still strange, so.. Just hide it in translated versions as we had before :)

File: addons/web_editor/controllers/main.py
Comment (younn-o) [reply addons/web_editor/controllers/main.py]: In the end, this was fixed with https://github.com/odoo/odoo/commit/f05adbc8b59c67cbec4847687a1ee07e620c2a7a

**Comment by qsm-odoo** [comment]
@rdeodoo Did not understand (~~read~~) everything you said but yes: I would remove that menu dialog in translate mode, it's probably really confusing and not worth code to support.

**Comment by pierre-pde** [comment]
> @rdeodoo Did not understand (~read~) everything you said but yes: I would remove that menu dialog in translate mode, it's probably really confusing and not worth code to support.

Agreed :)

**Comment by rdeodoo** [comment]
Not sure which part you did not understand but since you are on the same page it's fine ;-)

**Comment by younn-o** [comment]
@qsm-odoo 4 small fixes, ready for review :)

File: addons/website/static/src/services/website_custom_menus.js
Comment (qsm-odoo) [inline addons/website/static/src/services/website_custom_menus.js:57]: Commit message is confusing, it says that the edit menu dialog is displayed in the user's language instead of the displayed website's language... but the only fix here is about hiding it when not displaying the website in its default language (as we chose). But does that mean that the edit menu dialog is still displayed in the user's language rather than website language ? :thinking: I tried to test... but apparently changing backend language does not work anymore -_-

Also, with this commit, the edit menu is properly hidden in translated websites... but not in translate mode :/ Was it like that before? It's even more important to fix it there as after you use, it reloads the iframe and breaks the translate mode.

File: addons/web_editor/static/src/components/media_dialog/media_dialog.js
Comment (qsm-odoo) [inline addons/web_editor/static/src/components/media_dialog/media_dialog.js:70]: As discussued in https://github.com/odoo/odoo/pull/100253, this commit should be removed right?

File: addons/website/static/src/services/website_service.js
Comment (qsm-odoo) [inline addons/website/static/src/services/website_service.js:125]: I guess it's ok but is not there a better thing to do? relying on a class on #wrapwrap in the iframe seems fragile, can't the localization service be used somehow?

File: addons/website/static/src/client_actions/website_preview/website_preview.js
Comment (qsm-odoo) [inline addons/website/static/src/client_actions/website_preview/website_preview.js:87]: I wonder if this commit is actually something we want? I would not mind the title still be prefixed by "Odoo -", actually indicating you are in the backend. Did somebody ask for this to be changed? If yes, then maybe it should be entirely consistent and also fix the facat the favicon of the backend is used? :thinking:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/web_editor/static/src/components/media_dialog/media_dialog.js
Comment (younn-o) [reply addons/web_editor/static/src/components/media_dialog/media_dialog.js:70]: Yes

File: addons/website/static/src/client_actions/website_preview/website_preview.js
Comment (qsm-odoo) [reply addons/website/static/src/client_actions/website_preview/website_preview.js:87]: Let's let @rdeodoo decide then but for:
- Either we use favicon of the website + no "Odoo -"
- Or we leave as it is right now

But not a mix of both...

File: addons/website/static/src/services/website_custom_menus.js
Comment (younn-o) [reply addons/website/static/src/services/website_custom_menus.js:57]: I updated the commit to always display the menus from the "Edit Menu" in the website's language, and hide that menu for a translatable website.
For the popover shown in translate mode, it was a bug coming from https://github.com/odoo/odoo/commit/df1869153a90898bad3e0b61e5fc43a9ed3d59c9

File: addons/website/static/src/client_actions/website_preview/website_preview.js
Comment (rdeodoo) [reply addons/website/static/src/client_actions/website_preview/website_preview.js:95]: IMHO it just looks buggy when you are optimizing the SEO of you page, you may wonder why is there an "Odoo -" prefixing the tab/title of your page and how to get rid of it.
image

I agree it might makes sense to have "Odoo -" prefix to illustrate we are in the backend but:
1. It is not needed, we have the backend navbar, it is already easy to figure
2. It would be coming with the problem that when optimizing SEO, it is not perfect (as explained above)

About the favicon, I don't mind either way really, as long as this prefix is gone, unless someone disagrees.
Let's change the favicon too then I guess :+1:

File: addons/website/static/src/services/website_service.js
Comment (qsm-odoo) [reply addons/website/static/src/services/website_service.js:129]: Let's agree to keep this, if not easy to access inline services

File: addons/website/static/src/client_actions/website_preview/website_preview.js
Comment (younn-o) [reply addons/website/static/src/client_actions/website_preview/website_preview.js]: Changed according to that :)
It should be ready for review

**Comment by younn-o** [comment]
@qsm-odoo This PR is ready for review :)

File: addons/website/static/src/components/dialog/seo.js
Comment (qsm-odoo) [inline addons/website/static/src/components/dialog/seo.js:273]: Cool... but then I hit the cancel button of the SEO dialog and the entered title is not discarded... :grimacing:

File: addons/website/views/website_templates.xml
Comment (qsm-odoo) [inline addons/website/views/website_templates.xml:1867]: I don't think the commit message explanation is right, neither is the fix:

- I don't think BS4 had a h-0 class, we just extended BS4 ones to have it
- You should check why that extension was removed and probably restore it instead

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@younn-o Probably easy fixes to do (ping me again once you adapted)

File: addons/website/views/website_templates.xml
Comment (qsm-odoo) [reply addons/website/views/website_templates.xml:1867]: Got curious: [it's this code](https://github.com/odoo/odoo/commit/fe4f1efb70c0a9c4ebcb2b51fe090c4731c399b4#diff-23677f06a9aaf2e3395e957cfb9580ce6fcc9aba2635b560fe8f15fa27707313R17) that has not been adapted to extend BS5 utilities the right way (cc @adr-odoo just so you know)

File: addons/website/static/src/client_actions/website_preview/website_preview.js
Comment (younn-o) [reply addons/website/static/src/client_actions/website_preview/website_preview.js:119]: I did not check deep enough, I wanted to replace the backend icon (and reverting it back to what it was), only if it was different from the frontend icon, but:
- There should have been another test here before replacing the backend icon: `if (frontendIconEl && frontendIconEl.href !== backendIconEl.href)`
- In the end, they are always different (`/web/image/website/...` !== `/web/image/res.company/...`)

... so tldr; no

File: addons/website/static/src/js/editor/wysiwyg.js
Comment (younn-o) [reply addons/website/static/src/js/editor/wysiwyg.js:276]: This override was introduced in 16.0, to listen clicks on both documents (the snippets one and the iframe's one)

File: addons/website/static/src/client_actions/website_preview/website_preview.js
Comment (qsm-odoo) [reply addons/website/static/src/client_actions/website_preview/website_preview.js:106]: With that kind of logic, you can obfuscate all lines of the code. If later we need more values from `backendIconEl`... we'll change the code to the most readable version which allows to extract values. Meanwhile I don't see the point in using an unreadable version to read one attribute :/ You really don't agree that `const backendIconHref = backendIconEl.href` is more readable?

Comment (younn-o) [reply addons/website/static/src/client_actions/website_preview/website_preview.js:106]: No, I don't agree, I think it's appropriate to use the es6 destructuring assignment here, but I'll follow your call :+1:

File: addons/website/views/website_templates.xml
Comment (younn-o) [reply addons/website/views/website_templates.xml]: I removed this commit from this branch, and moved it here: https://github.com/odoo/odoo/pull/101541

File: addons/website/static/src/client_actions/website_preview/website_preview.js
Comment (qsm-odoo) [reply addons/website/static/src/client_actions/website_preview/website_preview.js]: Ok, then I'll guess I'll try not to make that kind of feedback anymore. But I really really don't understand :grimacing: (but I don't like that form even with multiple values TBH, I feel it should be the other way around (`backendIconHref: href`) so it can be read `myVariable = theValue` = "create variables whose value is this attribute from this object"). But there are 46 occurences exactly like this in the codebase, so...

I guess many people can read

"create a variable" "with the href property" "to be named backendIconHref" "to be set" "using the backendIconEl to read the property"

The same way as

"create a variable" "named backendIconHref" "to be set" "from the backendIconEl variable" "reading the href property"

Comment (younn-o) [reply addons/website/static/src/client_actions/website_preview/website_preview.js]: I guess it's a matter of habit, I got used to use destructuring reading it as you described it, working with react, most of the time component will be defined as:
```js
const Greetings = ({ name, lastName }) => {
  return <p>Hello {name} {lastName}<p>
}
```
instead of 
```js
const Greetings = (props) => {
  return <p>Hello {props.name} {props.lastName}<p>
}
```

Comment (qsm-odoo) [reply addons/website/static/src/client_actions/website_preview/website_preview.js]: It's not the same form here. I don't mind destructuring with `const {name} = props` because it that case, it reads exactly as it is (even more true with function params ("take these properties from the object given as argument")). It's really the form `const {name: myVariable} = props` I have issues with.

**Comment by younn-o** [comment]
@qsm-odoo This PR is ready for review

**Comment by younn-o** [comment]
@qsm-odoo This PR is ready for review

**Review by qsm-odoo — APPROVED** [review]
@robodoo rebase-ff r+

**Pull request closed without merging**
